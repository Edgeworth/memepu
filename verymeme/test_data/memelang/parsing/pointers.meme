fn main() i8 {
  val := i8(2);
  ptr : *i8 = &val;
  pptr : **i8 = &ptr;

  cval := i8(2);
  cptr_ci8 : const* const i8 = &cval;
  cptr_cptr_ci8 : const* const* const i8 = &cptr_ci8;

  array_cptr_ci8 : [42] const* const i8 = {};
  cptr_array_ci8 : const* [42] const i8 = 0;

  multidim_array : [40][30]i8 = {};
  cmultidim_array : [40][30]const i8 = {};
}
