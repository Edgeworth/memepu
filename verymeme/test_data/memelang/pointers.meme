// status: ok
// input: ''
// output: ''
// returns: 0
fn main() i8 {
  val := i8{2};
  ptr : *i8 = &val;
  pptr : **i8 = &ptr;

  cval := i8{2};
  cptr_ci8 : const* const i8 = &cval;
  cptr_cptr_ci8 : const* const* const i8 = &cptr_ci8;

  array_cptr_ci8 : [42] const* const i8 = [42]const* const i8{};
  array_cptr_ci8_deduce1 := [42]const* const i8{};
  array_cptr_ci8_deduce2 : [42] const* const i8 = {};
  cptr_array_ci8 : const* [42] const i8 = *[42]i8{0};
  cptr_array_ci8_deduce1 := *[42]i8{0};
  cptr_array_ci8_deduce2 : const* [42] const i8 = {};

  multidim_array : [40][30]i8 = [40][30]i8{};
  multidim_array_deduce1 := [40][30]i8{};
  multidim_array_deduce2 : [40][30]i8 = {};
  cmultidim_array : [40][30]const i8 = [40][30]const i8{};
  cmultidim_array_deduce1 := [40][30]const i8{};
  cmultidim_array_deduce2 : [40][30]const i8 = {};
  ret 0;
}
