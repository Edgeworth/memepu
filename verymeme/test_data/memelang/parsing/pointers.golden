File
 FnDefn
  FnSig
   Typename(main)
   Type(i8)
  StmtBlk
   Var
    VarDecl
     VarRef(val)
     Type(i8)
    IntLit(2)
   Var
    VarDecl
     VarRef(ptr)
     Type(i8)
      Qualifier(const: 0 ptr: 1, array: 0)
    Op(UNARY_ADDR)
     VarRef(val)
   Var
    VarDecl
     VarRef(pptr)
     Type(i8)
      Qualifier(const: 0 ptr: 1, array: 0)
      Qualifier(const: 0 ptr: 1, array: 0)
    Op(UNARY_ADDR)
     VarRef(ptr)
   Var
    VarDecl
     VarRef(cval)
     Type(const i8)
    IntLit(2)
   Var
    VarDecl
     VarRef(cptr_ci8)
     Type(const i8)
      Qualifier(const: 1 ptr: 1, array: 0)
    Op(UNARY_ADDR)
     VarRef(cval)
   Var
    VarDecl
     VarRef(cptr_cptr_ci8)
     Type(const i8)
      Qualifier(const: 1 ptr: 1, array: 0)
      Qualifier(const: 1 ptr: 1, array: 0)
    Op(UNARY_ADDR)
     VarRef(cptr_ci8)
   Var
    VarDecl
     VarRef(array_cptr_ci8)
     Type(const i8)
      Qualifier(const: 0 ptr: 0, array: 42)
      Qualifier(const: 1 ptr: 1, array: 0)
    CompoundLit
   Var
    VarDecl
     VarRef(cptr_array_ci8)
     Type(const i8)
      Qualifier(const: 1 ptr: 1, array: 0)
      Qualifier(const: 0 ptr: 0, array: 42)
    IntLit(0)
   Var
    VarDecl
     VarRef(multidim_array)
     Type(i8)
      Qualifier(const: 0 ptr: 0, array: 40)
      Qualifier(const: 0 ptr: 0, array: 30)
    CompoundLit
   Var
    VarDecl
     VarRef(cmultidim_array)
     Type(const i8)
      Qualifier(const: 0 ptr: 0, array: 40)
      Qualifier(const: 0 ptr: 0, array: 30)
    CompoundLit
